serrors/serrors.go

```diff
@@ -17,38 +23,68 @@ func (e *Error) Unwrap() error {
 
 // Error returns the string representation of the error
 func (e *Error) Error() string {
-	return e.error.Error()
+	var elems []string
+	keys := lo.Keys(e.keysAndValues)
+	sort.Strings(keys) // sort keys so we always get a consistent ordering
+	for _, k := range keys {
+		v := e.keysAndValues[k]
+		elems = append(elems, fmt.Sprintf("%s=%v", k, v))
+	}
+	return fmt.Sprintf("%s (%s)", e.error.Error(), strings.Join(elems, ", "))
 }
 
 // WithValues injects additional structured keys and values into the error
 func (e *Error) WithValues(keysAndValues ...any) *Error {
-	e.keysAndValues = append(e.keysAndValues, keysAndValues...)
+	for i := 0; i < len(keysAndValues); i += 2 {
+		e.keysAndValues[keysAndValues[i].(string)] = keysAndValues[i+1]
+	}
 	return e
 }
 
 // Wrap wraps and existing error with additional structured keys and values
 func Wrap(err error, keysAndValues ...any) error {
+	m := map[string]any{}
 	if len(keysAndValues)%2 != 0 {
 		panic("keysAndValues must have an even number of elements")
 	}
 	for i := 0; i < len(keysAndValues); i += 2 {
 		if _, ok := keysAndValues[i].(string); !ok {
 			panic("keys must be strings")
 		}
+		m[keysAndValues[i].(string)] = keysAndValues[i+1]
 	}
-	return &Error{error: err, keysAndValues: keysAndValues}
+	return &Error{error: err, keysAndValues: m}
 }
 
 // UnwrapValues returns a combined set of keys and values from every wrapped error
-func UnwrapValues(err error) (values []any) {
+func UnwrapValues(err error) (res []any) {
+	values := map[string][]any{}
 	for err != nil {
-		if v, ok := err.(*Error); ok {
-			values = append(values, v.keysAndValues...)
+		for _, elem := range multierr.Errors(err) {
+			if e, ok := elem.(*Error); ok {
+				for k, v := range e.keysAndValues {
+					if _, mOk := values[k]; mOk {
+						values[k] = append(values[k], v)
+					} else {
+						values[k] = []any{v}
+					}
+				}
+			}
 		}
 		err = errors.Unwrap(err)
 	}
 	if len(values) == 0 {
 		return nil
 	}
-	return values
+	keys := lo.Keys(values)
+	sort.Strings(keys) // sort keys so we always get a consistent ordering
+	for _, k := range keys {
+		v := values[k]
+		if len(v) == 1 {
+			res = append(res, k, v[0])
+		} else {
+			res = append(res, fmt.Sprintf("%ss", k), v)
```

Comment:
To match the previous behavior, should we have duplicate pairs of keys and values rather than a single key followed by multiple values? Let's say we have a key (`foo`) with multiple values from different errors (`bar` and `baz`). Right now the result would be `[]any{"foos", []any{"bar", "baz"}}`, but do we want it to actually be `[]any{"foo", "bar", "foo", "baz"}`? I think what you have currently makes sense given the context I know it's used in, but wanted to sanity check.

Reply:
I'd rather than the top one personally, that's going to lead to a much more readable structured log